netutils/interface.py

```diff
@@ -172,3 +176,179 @@ def abbreviated_interface_name(interface, addl_name_map=None, addl_reverse_map=N
         raise ValueError(f"Verify interface on and no match found for {interface}")
     # If abbreviated name lookup fails, return original name
     return interface
+
+
+@total_ordering
+class CharacterClass(ABC):
+    """CharacterClass embodies the state needed to sort interfaces."""
+
+    def __init__(self, val: str, terminal: bool = False) -> None:  # noqa: D107
+        self.val = val
+        self._terminal = terminal
+        super().__init__()
+
+    @abstractmethod
+    def __lt__(self, other) -> bool:  # noqa: D105
+        ...
+
+    def __eq__(self, other) -> bool:  # noqa: D105
+        return self.weight == other.weight and self.val == other.val
+
+    @abstractproperty
+    def weight(self) -> int:
+        """Weight property."""
+        ...
+
+    @property
+    def terminal(self):
+        """Flag whether a node is terminal."""
+        return self._terminal
+
+    @terminal.setter
+    def terminal(self, val: bool) -> None:
+        """This is a one-way switch to prevent overwriting a terminal node."""
+        if not self._terminal:
+            self._terminal = val
+
+    def __str__(self) -> str:  # noqa: D105
+        return str(self.val)
+
+    def __hash__(self) -> int:  # noqa: D105
+        return self.val.__hash__()
+
+
+class CCString(CharacterClass):
+    """Strings are sorted lexicographically."""
+
+    def __lt__(self, other) -> bool:  # noqa: D105
+        return self.weight < other.weight or self.val < other.val
+
+    def __repr__(self) -> str:  # noqa: D105
+        return f'CCString("{self.val}", {self.terminal})'
+
+    @property
+    def weight(self) -> int:  # noqa: D107,D102
+        return 10
+
+
+class CCInt(CharacterClass):
+    """Ints must be sorted canonically because '11' < '5'."""
+
+    def __lt__(self, other) -> bool:  # noqa: D105
+        return self.weight < other.weight or int(self.val) < int(other.val)
+
+    def __repr__(self) -> str:  # noqa: D105
+        return f"CCInt({self.val}, {self.terminal})"
+
+    @property
+    def weight(self) -> int:  # noqa: D107,D102
+        return 20
+
+
+class CCSeparator(CharacterClass):
+    """Separators require custom logic, so we sort them by arbitrary weight."""
+
+    weights: t.Dict[str, int] = {".": 10, "/": 20}
+
+    def __lt__(self, other) -> bool:  # noqa: D105
+        return self.weight < other.weight or self.weights.get(self.val, 0) < self.weights.get(other.val, 0)
+
+    def __repr__(self) -> str:  # noqa: D105
+        return f'CCSeparator("{self.val}", {self.terminal})'
+
+    @property
+    def weight(self) -> int:  # noqa: D102
+        return 30
+
+
+def _CCfail(*args):  # pylint: disable=C0103
+    """Helper to raise an exception on a bad character match."""
+    raise ValueError(f"unknown character '{args[0][0]}'.")
+
+
+def split_interface_tuple(interface: str) -> t.Tuple[CharacterClass, ...]:
```

Comment:
Can we mark split_interface_tuple and insert_nodes as private? I rather be explicit that these are not supported functions and subject to change.

Reply:
Yes, no problem.